rp2040-hal/src/i2c/controller.rs

```diff
@@ -100,89 +115,136 @@ where
         Self {
             i2c,
             pins: (sda_pin, scl_pin),
-            mode: PhantomData,
+            mode: Controller {},
         }
     }
 }
-impl<T: Deref<Target = Block>, PINS> I2C<T, PINS, Controller> {
-    fn validate(
-        addr: u16,
-        opt_tx_empty: Option<bool>,
-        opt_rx_empty: Option<bool>,
-    ) -> Result<(), Error> {
-        // validate tx parameters if present
-        if opt_tx_empty.unwrap_or(false) {
-            return Err(Error::InvalidWriteBufferLength);
-        }
-
-        // validate rx parameters if present
-        if opt_rx_empty.unwrap_or(false) {
-            return Err(Error::InvalidReadBufferLength);
-        }
 
-        // validate address
-        if addr >= 0x80 {
-            Err(Error::AddressOutOfRange(addr))
-        } else if i2c_reserved_addr(addr) {
-            Err(Error::AddressReserved(addr))
-        } else {
+impl<T: Deref<Target = Block>, PINS> I2C<T, PINS, Controller> {
+    fn validate_buffer<U>(
+        &mut self,
+        first: bool,
+        buf: &mut core::iter::Peekable<U>,
+        err: Error,
+    ) -> Result<(), Error>
+    where
+        U: Iterator,
+    {
+        if buf.peek().is_some() {
             Ok(())
+        } else {
+            if !first {
+                self.abort();
+            }
+            Err(err)
         }
     }
 
-    fn setup(&mut self, addr: u16) {
+    fn setup<A: ValidAddress>(&mut self, addr: A) -> Result<(), Error> {
+        addr.is_valid()?;
+
         self.i2c.ic_enable.write(|w| w.enable().disabled());
+        self.i2c
+            .ic_con
+            .modify(|_, w| w.ic_10bitaddr_master().variant(A::BIT_ADDR_M));
+
+        let addr = addr.into();
         self.i2c.ic_tar.write(|w| unsafe { w.ic_tar().bits(addr) });
         self.i2c.ic_enable.write(|w| w.enable().enabled());
+        Ok(())
     }
 
-    fn read_and_clear_abort_reason(&mut self) -> Option<u32> {
+    #[inline]
+    fn read_and_clear_abort_reason(&mut self) -> Result<(), Error> {
```

Comment:
What about `check_and_clear_abort_reason`?
Making it obvious that the abort reason is cleared isn't wrong, so it can be kept in the name. And `check_` doesn't imply that getting an abort reason is a successful result, in the same way as `read_` does.